### DISCUSSION TITLE
bx_dbg_read_linear: physical memory read error on chapter 2 First Steps

### DISCUSSION THREAD

**Discussion by arun-mani-j** [OrdoFlammae/littleosbook-src#20] [Q&A] [answered]
Hii all.

I followed the exact steps and copied the code verbatim under https://ordoflammae.github.io/littleosbook/#hello-cafebabe.
The only change I did is, using `sdl2` instead of `sdl` in [Bochs' configuration](https://ordoflammae.github.io/littleosbook/#running-bochs).

This is what I get when I run `bochs -f bochsrc.txt -q`:

```c
00000000000i[      ] LTDL_LIBRARY_PATH not set. using compile time default '/usr/lib/x86_64-linux-gnu/bochs/plugins'
========================================================================
                        Bochs x86 Emulator 2.8
             Built from GitHub snapshot on March 10, 2024
                Timestamp: Sun Mar 10 08:00:00 CET 2024
========================================================================
00000000000i[      ] BXSHARE not set. using compile time default '/usr/share/bochs'
00000000000i[      ] lt_dlhandle is 0x55da8b081160
00000000000i[PLUGIN] loaded plugin libbx_parallel.so
00000000000i[      ] lt_dlhandle is 0x55da8b082e00
00000000000i[PLUGIN] loaded plugin libbx_speaker.so
00000000000i[      ] lt_dlhandle is 0x55da8b083ee0
00000000000i[PLUGIN] loaded plugin libbx_biosdev.so
00000000000i[      ] lt_dlhandle is 0x55da8b084920
00000000000i[PLUGIN] loaded plugin libbx_gameport.so
00000000000i[      ] lt_dlhandle is 0x55da8b0851b0
00000000000i[PLUGIN] loaded plugin libbx_extfpuirq.so
00000000000i[      ] lt_dlhandle is 0x55da8b085a10
00000000000i[PLUGIN] loaded plugin libbx_iodebug.so
00000000000i[      ] lt_dlhandle is 0x55da8b086290
00000000000i[PLUGIN] loaded plugin libbx_serial.so
00000000000i[      ] lt_dlhandle is 0x55da8b08a490
00000000000i[PLUGIN] loaded plugin libbx_unmapped.so
00000000000i[      ] reading configuration from bochsrc.txt
00000000000i[      ] lt_dlhandle is 0x55da8b08af50
00000000000i[PLUGIN] loaded plugin libbx_textconfig.so
00000000000i[      ] lt_dlhandle is 0x55da8b08b6e0
00000000000i[PLUGIN] loaded plugin libbx_sdl2_gui.so
00000000000i[      ] installing sdl2 module as the Bochs GUI
00000000000i[SDL2  ] maximum host resolution: x=1920 y=1080
00000000000i[      ] using log file bochslog.txt
Bochs internal debugger, type 'help' for help or 'c' to continue
Switching to CPU0
Next at t=0
(0) [0x0000fffffff0] f000:fff0 (unk. ctxt): jmpf 0xf000:e05b          ; ea5be000f0
<bochs:1> c
bx_dbg_read_linear: physical memory read error (phy=0x0000322f3331, lin=0x00000000322f3331)
Next at t=662802
(0) [0x0000fffffff0] f000:fff0 (unk. ctxt): jmpf 0xf000:e05b          ; ea5be000f0
<bochs:2> c
bx_dbg_read_linear: physical memory read error (phy=0x0000322f3331, lin=0x00000000322f3331)
Next at t=1325794
(0) [0x0000fffffff0] f000:fff0 (unk. ctxt): jmpf 0xf000:e05b          ; ea5be000f0
<bochs:3> c
bx_dbg_read_linear: physical memory read error (phy=0x0000322f3331, lin=0x00000000322f3331)
Next at t=1988672
(0) [0x0000fffffff0] f000:fff0 (unk. ctxt): jmpf 0xf000:e05b          ; ea5be000f0
<bochs:4> q
(0).[1988672] [0x0000fffffff0] f000:fff0 (unk. ctxt): jmpf 0xf000:e05b          ; ea5be000f0
```

This is my [bochslog.txt](https://github.com/user-attachments/files/17083572/bochslog.txt).

I'm on Debian GNU/Linux Trixie/Sid with Bochs 2.8.

What am I doing wrong, thanks in advance!

**Comment by arun-mani-j** [comment]
[chapter-2.zip](https://github.com/user-attachments/files/17098344/chapter-2.zip)

The files of concern as discussed in #18

**Comment by OrdoFlammae** [comment]
@arun-mani-j, I stepped through all the different steps, comparing my files to yours, and it appears that everything becomes different when you create your ISO file. All the binaries before that are identical. Additionally, when I run my ISO, it works perfectly. What is your grub setup? Any logs or warnings when running `grub-mkrescue -o os.iso iso`? My guess is that the issue originates there.

**Reply by arun-mani-j** [reply]
Thanks for offering to help. There are no errors or warnings, but this is the output:

```sh
$ grub-mkrescue -o os.iso iso
xorriso 1.5.6 : RockRidge filesystem manipulator, libburnia project.

Drive current: -outdev 'stdio:os.iso'
Media current: stdio file, overwriteable
Media status : is blank
Media summary: 0 sessions, 0 data blocks, 0 data,  182g free
Added to ISO image: directory '/'='/tmp/grub.OScW0X'
xorriso : UPDATE :     342 files added in 1 seconds
Added to ISO image: directory '/'='/home/arun-mani-j/Labs/Little OS/iso'
xorriso : UPDATE :     346 files added in 1 seconds
ISO image produced: 7147 sectors
Written to medium : 7147 sectors at LBA 0
Writing to 'stdio:os.iso' completed successfully.
```

Is it similar to what you get too?

**Reply by OrdoFlammae** [reply]
My logs are almost identical. When I got a chance, I mounted your ISO and mine to see what the difference between the two was. The first thing that stood out to me was that your ISO is missing all the `i386` files, only including `x86_64` files. I don't know entirely why this is, you'd probably have to do some probing into your GRUB installation, but it appears that your installation is missing some crucial pieces. Note that BOCHS emulates an IA32 system, not a 64-bit system, so I would guess that this is your problem.

Once you figure out what packages you need to install to fix this issue, please let me know and/or create a PR to add a note to the book for future users.

Your ISO's folders:
```
2.4M	test/boot/grub/fonts
5.0M	test/boot/grub/locale
2.0K	test/boot/grub/roms
2.8M	test/boot/grub/x86_64-efi
10M		test/boot/grub
10M		test/boot
202K	test/efi/boot
204K	test/efi
205K	test/System/Library/CoreServices
207K	test/System/Library
209K	test/System
2.0K	test/.disk
14M		test
```

My ISO's folders:
```
2.4M	test2/boot/grub/fonts
2.0M	test2/boot/grub/i386-efi
2.1M	test2/boot/grub/i386-pc
11M		test2/boot/grub/locale
2.0K	test2/boot/grub/roms
2.8M	test2/boot/grub/themes/starfield
2.8M	test2/boot/grub/themes
2.7M	test2/boot/grub/x86_64-efi
23M		test2/boot/grub
23M		test2/boot
498K	test2/efi/boot
500K	test2/efi
354K	test2/System/Library/CoreServices
356K	test2/System/Library
358K	test2/System
2.0K	test2/.disk
27M		test2
```

**Reply by arun-mani-j** [reply]
Thanks, I seem to be missing `grub-pc-bin` for the pc files and `grub-efi-ia32-bin` for the `i386` files.

This is my image contents now:
```
$ tree -L 3 /media/arun-mani-j/ISOIMAGE/
/media/arun-mani-j/ISOIMAGE/
├── boot
│   ├── grub
│   │   ├── fonts
│   │   ├── grub.cfg
│   │   ├── i386-efi
│   │   ├── i386-pc
│   │   ├── locale
│   │   ├── roms
│   │   └── x86_64-efi
│   └── kernel.elf
├── efi
│   └── boot
│       ├── boot.efi
│       ├── bootia32.efi
│       └── bootx64.efi
├── mach_kernel
└── System
    └── Library
        └── CoreServices
```

But running `bochs` still gives the physical memory read error. :(

These are the packages I have regarding `grub`:
```sh
$ apt list --installed | grep grub-
grub-common/testing,now 2.12-5 amd64 [installed]
grub-efi-amd64-bin/testing,now 2.12-5 amd64 [installed,automatic]
grub-efi-amd64-signed/testing,now 1+2.12+5 amd64 [installed,automatic]
grub-efi-amd64-unsigned/testing,now 2.12-5 amd64 [installed,automatic]
grub-efi-amd64/testing,now 2.12-5 amd64 [installed]
grub-efi-ia32-bin/testing,now 2.12-5 amd64 [installed]
grub-efi-ia32-unsigned/testing,now 2.12-5 amd64 [installed,automatic]
grub-pc-bin/testing,now 2.12-5 amd64 [installed]
```
And my `os.iso`:
[os.zip](https://github.com/user-attachments/files/17165649/os.zip)

Can you upload your ISO file so I can check if it is not something wrong with my `bochs`? Thanks!

**Reply by OrdoFlammae** [reply]
Yes, it does look like it's an issue with your bochs setup, I ran the ISO you attached and it worked perfectly. It's not exactly the same as my ISO, but I am wondering if GRUB just attaches every single file it has access to into the ISO.

I've attached my OS file here so you can test it, but I doubt you will get anything new.
[os.zip](https://github.com/user-attachments/files/17172156/os.zip)

**Reply by arun-mani-j** [reply, marked as answer]
I must have been wearing my lucky hat - I got so nervous, I tried the images available in Bochs' sample images and they also had the same issue. I reinstalled packages, changed memory limits but everything failed with same error.

Then I tried searching for the errors in web, though most of them were due to buggy programs... then I found this AskUbuntu post - https://askubuntu.com/questions/1521719/bochs-cant-access-bootloader-address-0x7c00-error-physical-memory-read-error and the answer by the OP themselves:

> aparrently because of a new update bochs pushed it kinda messed with the romimage on Ubuntu as a quick fix instead of using the latest i used the legacy bios so instead of:
> 
> ```
> romimage: file=/usr/share/bochs/BIOS-bochs-latest, address = 0xfffe0000
> ```
> 
> just put:
> 
> ```
> romimage: file=/usr/share/bochs/BIOS-bochs-legacy
> ```

Oh my, I tried it and it worked!!

This is what I have in my `bochslog.txt`:
```
00553148000p[SDL2  ] >>PANIC<< POWER button turned off.
00553148000i[CPU0  ] CPU is in protected mode (active)
00553148000i[CPU0  ] CS.mode = 32 bit
00553148000i[CPU0  ] SS.mode = 32 bit
00553148000i[CPU0  ] EFER   = 0x00000000
00553148000i[CPU0  ] | EAX=cafebabe  EBX=00010000  ECX=00000000  EDX=00000000
00553148000i[CPU0  ] | ESP=0007ff00  EBP=00000000  ESI=00000000  EDI=00000000
00553148000i[CPU0  ] | IOPL=0 id vip vif ac vm rf nt of df if tf sf ZF af PF cf
00553148000i[CPU0  ] | SEG sltr(index|ti|rpl)     base    limit G D
00553148000i[CPU0  ] |  CS:0010( 0002| 0|  0) 00000000 ffffffff 1 1
00553148000i[CPU0  ] |  DS:0018( 0003| 0|  0) 00000000 ffffffff 1 1
00553148000i[CPU0  ] |  SS:0018( 0003| 0|  0) 00000000 ffffffff 1 1
00553148000i[CPU0  ] |  ES:0018( 0003| 0|  0) 00000000 ffffffff 1 1
00553148000i[CPU0  ] |  FS:0018( 0003| 0|  0) 00000000 ffffffff 1 1
00553148000i[CPU0  ] |  GS:0018( 0003| 0|  0) 00000000 ffffffff 1 1
00553148000i[CPU0  ] | EIP=00100011 (00100011)
00553148000i[CPU0  ] | CR0=0x60000011 CR2=0x00000000
00553148000i[CPU0  ] | CR3=0x00000000 CR4=0x00000000
00553148000i[CMOS  ] Last time: 1727533210 tz=utc (Sat Sep 28 14:20:10   2024)
00553148000i[SIM   ] quit_sim called with exit code 1
```

Feels so good now!

The person asked the question just 2 months ago.

I tried downloading the RPM package from Bochs website itself - https://sourceforge.net/projects/bochs/files/bochs/2.7/. Then I extracted it and found the BIOS images it had.
I tried using the `BIOS-bochs-latest` from it and it worked fine.

So let me try reporting it to Debian.

Thanks a lot for helping through the process :smile: .
